### PR TITLE
Feature/spec filter 3.0

### PR DESCRIPTION
**Pull request by ralphdoe**
Spec Filter New Version
Classes, Methods, Variables Renaming.
New OpenApiSpecFitler Interface
AbsctractFilter implementation
Starting the new SpecFilter Implementation
Filters added:
 - OpenAPI
 - Operations
 - Parameters
 - PathItem
 - Schemas related with Parameters
 - Tests added

### CODE REVIEW

File: modules/swagger-core/src/main/java/io/swagger/core/filter/SpecFilter.java [hunk 1]
```diff
@@ -4,314 +4,120 @@
...
+                filteredPathItem.setHead(filterOperation(filter, pathItem.getHead(), resourcePath, HEAD, params, cookies, headers));
+                filteredPathItem.setOptions(filterOperation(filter, pathItem.getOptions(), resourcePath, OPTIONS, params, cookies, headers));
+                filteredPathItem.setDescription(pathItem.getDescription());
+                filteredPathItem.set$ref(pathItem.get$ref());
+                filteredPathItem.setSummary(pathItem.getSummary());
+                clonedPaths.addPathItem(resourcePath, filteredPathItem);
             }
+            clone.paths(clonedPaths);
+            clone.setComponents(openAPI.getComponents());
+            clone.setSecurity(openAPI.getSecurity());
```
Comment (frantuma) [inline modules/swagger-core/src/main/java/io/swagger/core/filter/SpecFilter.java]: these should be called outside the loop

Comment (frantuma) [reply modules/swagger-core/src/main/java/io/swagger/core/filter/SpecFilter.java]: also other elements are not cloned? e.g. extensions, externalDocs, servers

File: modules/swagger-core/src/main/java/io/swagger/core/filter/SpecFilter.java [hunk 2]
```diff
@@ -4,314 +4,120 @@
...
+            if (filteredOperation.isPresent()) {
+                List<Parameter> filteredParameters = new ArrayList<>();
+                Operation filteredOperationGet = filteredOperation.get();
+                List<Parameter> parameters = filteredOperationGet.getParameters();
+                if (parameters != null) {
+                    for (Parameter parameter : parameters) {
+                        Parameter filteredParameter = filterParameter(filter, operation, parameter, resourcePath, key, params, cookies, headers);
+                        if (filteredParameter != null) {
+                            filteredParameter.setSchema(filterProperty(filter, parameter.getSchema(), resourcePath, key, params, cookies, headers));
+                            filteredParameters.add(filteredParameter);
```
Comment (frantuma) [inline modules/swagger-core/src/main/java/io/swagger/core/filter/SpecFilter.java]: see also general review comment, filterProperty should be used for schemas inside components also

File: modules/swagger-core/src/main/java/io/swagger/core/filter/SpecFilter.java (hunk 1 above)
Comment (ralphdoe) [reply modules/swagger-core/src/main/java/io/swagger/core/filter/SpecFilter.java]: Done!

**Comment by frantuma** [comment]
included in #2452

**Pull request closed without merging**